lib/bson/decimal128.rb

```diff
@@ -0,0 +1,315 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'bigdecimal'
+require 'bson/decimal128/builder'
+
+module BSON
+
+  class Decimal128
+
+    # A Decimal128 is type 0x13 in the BSON spec.
+    #
+    # @since 4.2.0
+    BSON_TYPE = 19.chr.force_encoding(BINARY).freeze
+
+    # Exponent offset.
+    #
+    # @since 4.2.0
+    EXPONENT_OFFSET = 6176.freeze
+
+    # Minimum exponent.
+    #
+    # @since 4.2.0
+    MIN_EXPONENT = -6176.freeze
+
+    # Maximum exponent.
+    #
+    # @since 4.2.0
+    MAX_EXPONENT = 6111.freeze
+
+    # Maximum digits of precision.
+    #
+    # @since 4.2.0
+    MAX_DIGITS_OF_PRECISION = 34.freeze
+
+    # Key for this type when converted to extended json.
+    #
+    # @since 4.2.0
+    EXTENDED_JSON_KEY = "$numberDecimal".freeze
+
+    # The native type to which this object can be converted.
+    #
+    # @since 4.2.0
+    NATIVE_TYPE = BigDecimal.freeze
+
+    # Get the Decimal128 as JSON hash data.
+    #
+    # @example Get the Decimal128 as a JSON hash.
+    #   decimal.as_json
+    #
+    # @return [ Hash ] The number as a JSON hash.
+    #
+    # @since 4.2.0
+    def as_json(*args)
+      { EXTENDED_JSON_KEY => to_s }
+    end
+
+    # Check equality of the decimal128 object with another object.
+    #
+    # @example Check if the decimal128 object is equal to the other.
+    #   decimal == other
+    #
+    # @param [ Object ] other The object to check against.
+    #
+    # @return [ true, false ] If the objects are equal.
+    #
+    # @since 4.2.0
+    def ==(other)
+      return false unless other.is_a?(Decimal128)
+      @high == other.instance_variable_get(:@high) &&
+          @low == other.instance_variable_get(:@low)
+    end
+    alias :eql? :==
+
+    # Create a new Decimal128 from a Integer, Float, Rational, BigDecimal, or String.
+    #
+    # @example Create a Decimal128 from an Integer.
+    #   Decimal128.new(100)
+    #
+    # @example Create a Decimal128 from an Float.
+    #   Decimal128.new(1.45, 3)
+    #
+    # @example Create a Decimal128 from an Rational.
+    #   Decimal128.new(Rational(2, 3), 2)
+    #
+    # @example Create a Decimal128 from an BigDecimal.
+    #   Decimal128.new(BigDecimal.new("2.45"))
+    #
+    # @example Create a Decimal128 from an String.
+    #   Decimal128.new("2.45")
+    #
+    # @param [ Integer, Float, Rational, BigDecimal, String ] initial The initial value to use for
+    #   instantiating a Decimal128.
+    # @param [ Fixnum ] digits The number of significant digits, as a Fixnum.
+    #   If omitted or 0, the number of significant digits is determined from the initial value.
+    #
+    # @since 4.2.0
+    def initialize(initial, digits = nil)
+      if initial.is_a?(String)
+        set_bits(*Builder::FromString.new(initial).bits)
+      else
+        big_decimal = BigDecimal.new(*[initial, digits].compact)
```

Comment:
@durran I've updated the #initialize method to create a Decimal128 directly from an Integer and Float, instead of instantiating a BigDecimal first. Benchmarking showed that this was significantly faster.